### ISSUE TITLE
Don't use the 'legacy' security realm

### ISSUE THREAD

**Issue by daniel-beck** [jenkinsci/helm-charts#20]
I spent a few hours helping a colleague with his Jenkins instance recently that didn't want to let him authenticate using username/API token, complaining about a missing CSRF token/crumb, which shouldn't be needed since [Jenkins 2.96](https://www.jenkins.io/changelog-old/#v2.96).

Turns out this chart configures the legacy security realm which is basically untested, and completely ignored for new development.

https://github.com/jenkinsci/helm-charts/blob/8ead78964ba900f610e5b7f86f6e6bcc1f077620/charts/jenkins/values.yaml#L93

https://github.com/jenkinsci/helm-charts/blob/8ead78964ba900f610e5b7f86f6e6bcc1f077620/charts/jenkins/values.yaml#L285

It probably shouldn't declare a security realm whose name already recommends against its use.

**Comment by jglick** [comment]
I suppose it would be helpful if there were a security realm for Jenkins which delegates to the Kubernetes cluster’s authentication system, but as far as I can tell [there is none](https://www.jenkins.io/doc/developer/extensions/jenkins-core/#securityrealm). The closest thing I can find is [this](https://github.com/jenkinsci/kubesphere-extension-plugin#3-basic-header-auth-with-kubesphere-iam). Anyway for a default realm to get people started we generally recommend `HudsonPrivateSecurityRealm`, or simply let the GUI setup wizard run so as to get an initial `admin` user.

**Comment by torstenwalter** [comment]
For production use one probably wants to change both `authorizationStrategy` and `securityRealm`:

https://github.com/jenkinsci/helm-charts/blob/8ead78964ba900f610e5b7f86f6e6bcc1f077620/charts/jenkins/templates/_helpers.tpl#L77-L84

XML configuration is something we want to get rid of completely see #10.

On the other hand I don't want to have a default configuration which uses no authentication at all. 
I think there is nothing like a Kubernetes authentication as that largely dependends on you setup. Could be OIDC or completely out of you control. So sometimes you just want to use good old LDAP or Active Directory authentication.

What would be a reasonable default here instead of `legacy`?

**Comment by jglick** [comment]
> What would be a reasonable default here instead of `legacy`?

As mentioned above, specify nothing, but let the setup wizard run. It will offer an initial `admin` account with a randomized password, and set the authorization strategy to let authenticated users do anything and anonymous users are forced to log in.

**Comment by torstenwalter** [comment]
I like the approach. The downside of the setup wizard is that it requires manual actions and is therefore not completely automated.

On the other hand it would be great to get rid of the secret which contains the admin password.

**Comment by daniel-beck** [comment]
JCasC can pre-configure a security realm with users defined, that's probably a reasonable approach to fully automate: https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/demos/embedded-userdatabase/README.md

**Comment by jglick** [comment]
> pre-configure a security realm with users defined

Right but then you have to hard-code an admin password, right?

**Comment by timja** [comment]
@jglick by default the chart auto generates a password IIRC

you can also specify it yourself and at least with jcasc not sure about this chart you can provide a pre-hashed password.

and providing secrets to helm values isn't normally an issue, sops, sealed-secrets, git-crypt etc are some of the many solutions for hiding secrets values inputs

**Comment by torstenwalter** [comment]
I can shed some light on how it's currently implemented:

Admin credentials can be configured via value file:
https://github.com/jenkinsci/helm-charts/blob/bdbc999ceea93512169feab6dfa74aaaad8f6704/charts/jenkins/values.yaml#L103-L108
The admin password is empty by default. In that case it will be auto generated. It is possible to provide a password via values file, but I rather recommend to use an existing secret instead as I believe secret values should not be passed via helm values.

If the case an existing secret is not provided this template renders a secret:
https://github.com/jenkinsci/helm-charts/blob/bdbc999ceea93512169feab6dfa74aaaad8f6704/charts/jenkins/templates/secret.yaml#L1-L22
That's also the place where the random password is generated (`jenkins-admin-password: {{ randAlphaNum 10 | b64enc | quote }}`)

Username and password are then exposed as environment variables to the Jenkins container
https://github.com/jenkinsci/helm-charts/blob/bdbc999ceea93512169feab6dfa74aaaad8f6704/charts/jenkins/templates/jenkins-master-deployment.yaml#L194-L205

and passed as args to Jeknins
https://github.com/jenkinsci/helm-charts/blob/bdbc999ceea93512169feab6dfa74aaaad8f6704/charts/jenkins/templates/jenkins-master-deployment.yaml#L175

**Comment by hferentschik** [comment]
@daniel-beck,

> JCasC can pre-configure a security realm with users defined, that's probably a reasonable approach to fully automate: https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/demos/embedded-userdatabase/README.md

Thanks that actually worked for my use case. I ended up using the following values.yaml:

```yaml
master:
  JCasC:
    configScripts:
      security: |
        jenkins:
          securityRealm:
            local:
              allowsSignup: false
              users:
                - id: "admin"
                  password: "admin"
          authorizationStrategy: loggedInUsersCanDoAnything
```

That works fine and using this setup the REST API is working fine :-)

The status message one gets during the Helm install is confusing though:

```bash
helm install jenkins jenkinsci/jenkins -f values.yaml
NAME: jenkins
LAST DEPLOYED: Thu Sep 17 11:14:39 2020
NAMESPACE: jenkins
STATUS: deployed
REVISION: 1
NOTES:
1. Get your 'admin' user password by running:
  printf $(kubectl get secret --namespace jenkins jenkins -o jsonpath="{.data.jenkins-admin-password}" | base64 --decode);echo
2. Get the Jenkins URL to visit by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace jenkins -l "app.kubernetes.io/component=jenkins-master" -l "app.kubernetes.io/instance=jenkins" -o jsonpath="{.items[0].metadata.name}")
  echo http://127.0.0.1:8080
  kubectl --namespace jenkins port-forward $POD_NAME 8080:8080

3. Login with the password from step 1 and the username: admin

4. Use Jenkins Configuration as Code by specifying configScripts in your values.yaml file, see documentation: http:///configuration-as-code and examples: https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos

For more information on running Jenkins on Kubernetes, visit:
https://cloud.google.com/solutions/jenkins-on-container-engine
For more information about Jenkins Configuration as Code, visit:
https://jenkins.io/projects/jcasc/
```

It seems there is still the default _Secret_ created which one can read via `printf $(kubectl get secret --namespace jenkins jenkins -o jsonpath="{.data.jenkins-admin-password}" | base64 --decode);`, however, the actual password for logging in is in the JCasC script. Probably this cannot be easily solved, but maybe a hint/warning could be added to the message?

**Comment by daniel-beck** [comment]
Looks like it would be a straightforward PR changing https://github.com/jenkinsci/helm-charts/blob/master/charts/jenkins/templates/NOTES.txt

**Comment by hferentschik** [comment]
> Looks like it would be a straightforward PR changing https://github.com/jenkinsci/helm-charts/blob/master/charts/jenkins/templates/NOTES.txt

Just adding a message would be trivial, yes. Finding out whether the default security context gets overridden might be harder.

**Comment by torstenwalter** [comment]
I somehow missed this discussion.

It's great that we can configure admin credentials via JCasC. However we should not put the password directly in the config as secrets should not be stored in a ConfigMap.

Configuration as Code plugin allows to reference secrets. 
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/features/secrets.adoc

So a way forward could be to configure admin via JCasC and reference the generated secret.

@timja What do you think?

**Comment by timja** [comment]
Makes sense as long as it’s easy to disable, I would expect many users will be configuring SSO after any initial testing is done and don’t need this.

So this should have sensible defaults (not using legacy and creating an admin user) but also have a simple way to not have this run

**Comment by torstenwalter** [comment]
If I remember correctly then `useSecurity` flag is used to configure admin user. If we give it a better name and document it in readme then we could use that mechanism.

**Comment by timja** [comment]
> If I remember correctly then `useSecurity` flag is used to configure admin user. If we give it a better name and document it in readme then we could use that mechanism.

maybe just `security`, not using security with jenkins is just asking for trouble.

It likely maps to the old boolean in the jenkins web ui which was use security but that was removed months ago

**Comment by torstenwalter** [comment]
I meant the `master.useSecurity` boolean which controls [creation of the secret for the admin user](https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/templates/secret.yaml#L1) and [creation of the admin user via parameters](https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/templates/jenkins-master-deployment.yaml#L177-L179).

I think we need a better name for it like `controller.testAdminUser`. recommending to disable security (useSecurity: false) feels strange.

We could also render a warn message in NOTES if that setting is used to tell the user that this is just for getting started and they need to configure something better for production use.

Anyhow this would be something for the 3.0.0 release as it's a breaking change. I want to avoid resetting security configuration for instance which configured things manually without JCasC.

**Comment by jglick** [comment]
Tip: https://github.com/helm/charts/issues/5167#issuecomment-641558251

**Comment by daniel-beck** [comment]
stalebot is seriously the worst bot I have ever seen.

**Comment by torstenwalter** [comment]
Closed by #158

**Comment by torstenwalter** [comment]
> stalebot is seriously the worst bot I have ever seen.

@daniel-beck you are welcome to help as contributor to work on issues and create PRs which resolve them.
